presto-hive/src/main/java/io/prestosql/plugin/hive/orc/acid/DeletedRowsRegistry.java

```diff
@@ -0,0 +1,423 @@
+import static io.prestosql.plugin.hive.HiveErrorCode.HIVE_BAD_DATA;
+import static io.prestosql.plugin.hive.HiveErrorCode.HIVE_UNKNOWN_ERROR;
+import static org.apache.hadoop.hive.serde.serdeConstants.SERIALIZATION_LIB;
+
+public class DeletedRowsRegistry
+{
+    private final List<DeletedRowsLoader> loaders;
+    private final String queryId;
+    private final boolean deltedRowsPresent;
+
+    private int[] validPositions;
+    private AtomicReference<Set<RowId>> deletedRows = new AtomicReference();
+
+    // variables to track cache hit rate, Guava Cache.stats() is not useful because multiple threads loading same
+    // value get synchronised and only one of them load value but all of them update cache-miss count.
+    // We want to know how many times did we read from cache rather
+    private static final AtomicLong totalRequests = new AtomicLong();
+    private static final AtomicLong requestsNeedingLoad = new AtomicLong();
+
+    // For a query, DELETE_DELTA rows for a partition location will not change, they are cached in this cache
+    private static Cache<String, List<RowId>> cache; // cachekey = QueryId_PartitionLocation_FileName
+    private static Boolean lock = new Boolean(true);
+
+    private static Properties orcProperty = new Properties();
+    private static final List<HiveColumnHandle> ACID_ROW_ID_COLUMN_HANDLES = createAcidRowIdMetaColumns();
+    private static final int ROW_ID_SIZE = Long.BYTES * 2 + Integer.BYTES;
+    private static final List<DeletedRowsLoader> EMPTY_LOADERS = ImmutableList.of();
+    private static final CacheStats EMPTY_STATS = new AbstractCache.SimpleStatsCounter().snapshot();
+
+    private static Cache<String, List<LocatedFileStatus>> fileStatusCache;
+
+    static {
+        orcProperty.setProperty(SERIALIZATION_LIB, OrcSerde.class.getName());
+    }
+
+    public DeletedRowsRegistry(
+            Path splitPath,
+            OrcPageSourceFactory pageSourceFactory,
+            ConnectorSession session,
+            Configuration configuration,
+            DateTimeZone hiveStorageTimeZone,
+            HdfsEnvironment hdfsEnvironment,
+            DataSize cacheSize,
+            Duration cacheTTL,
+            Optional<DeleteDeltaLocations> deleteDeltaLocations)
+            throws ExecutionException
+    {
+        if (cache == null) {
+            synchronized (lock) {
+                if (cache == null) {
+                    cache = CacheBuilder.newBuilder()
+                            .expireAfterAccess(cacheTTL.toMillis(), TimeUnit.MILLISECONDS)
+                            .maximumWeight((int) cacheSize.toBytes())
+                            .weigher((Weigher<String, List<RowId>>) (key, value) -> {
+                                return ROW_ID_SIZE * value.size();
+                            })
+                            .recordStats()
+                            .build();
+
+                    fileStatusCache = CacheBuilder.newBuilder()
```

Comment:
Will fix the initialization part.

I had plans to extend cache usage across queries next. It will need some tweaking to drop queryInfo and add delete_delta id into the cache. Hence I would like to retain it like this to keep that change simpler.

Reply:
About the static init part, some parameters like ttl will not be available during static init hence using this way. I will separate the inits of two caches to prevent the issue you mentioned.